src/ra_server.erl

```diff
@@ -2451,16 +2479,34 @@ append_log_leader({CmdTag, _, _, _},
   when CmdTag == '$ra_join' orelse
        CmdTag == '$ra_leave' ->
     {not_appended, cluster_change_not_permitted, State};
+append_log_leader({'$ra_join', From, #{node := JoiningNode, voter := Voter}, ReplyMode},
+                  State = #{cluster := OldCluster}) ->
+    case OldCluster of
+        #{JoiningNode := #{voter := Voter}} ->
+            % already a member do nothing
+            % TODO: reply? If we don't reply the caller may block until timeout
+            {not_appended, already_member, State};
+        #{JoiningNode := Peer} ->
+            % Update member status.
+            Cluster = OldCluster#{JoiningNode => Peer#{voter => Voter}},
```

Comment:
Does this mean that voter demotion is also possible?

Reply:
Correct! Nothing in Ra does that yet, but my reasoning is that library users will appreciate the capability.